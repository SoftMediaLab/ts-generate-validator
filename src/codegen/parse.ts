import { parseStruct, Decorator, ImportNode } from 'ts-file-parser';
import * as fs from 'fs';
// import * as path from 'path';
// import { ArrayType, BasicType } from 'ts-file-parser';

export interface DecoratedClassFieldMetadata {
  name: string;
  type: any;
  optional: any;
  decorators: Decorator[];
}

export interface DecoratedClassMetadata {
  name: string;
  decorators: Decorator[];
  fields: DecoratedClassFieldMetadata[];
}

export interface InputFileMetadata {
  name: string;
  imports: string[];
  classes: DecoratedClassMetadata[];
}

export const parseInputFiles = (files: string[]): InputFileMetadata[] => {
  return files.map((file) => {
    console.log(`Reading ${file}...`);
    const content = fs.readFileSync(file, 'utf-8');
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const structure = parseStruct(content, {}, file);

    //if (file.match(/^.*common.ts$/)) {
    //console.log(JSON.stringify(structure, null, 4));
    //}

    return {
      name: structure.name.replace(/\/+/g, '.'),
      imports: buildImports(structure._imports),
      classes: []
    };
  });
};

const buildImports = (importNodes: ImportNode[]): string[] => {
  return importNodes.map((iNode) => {
    return `import { ${iNode.clauses.join(', ')} } from ${iNode.absPathNode.join('/')}`;
  });
};
