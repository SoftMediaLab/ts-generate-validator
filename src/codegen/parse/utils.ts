import { primitiveValidationTypes } from './../../validators/model';
import { ValidationType } from '../../validators/model';

export const getValidationTypeByTypeName = (typeName: string): ValidationType => {
  if (primitiveValidationTypes.includes(typeName as any)) {
    return <ValidationType>typeName;
  }

  const notSupportedTypes = ['null', 'undefined', 'symbol', 'function', 'object'];
  if (notSupportedTypes.includes(typeName)) {
    return ValidationType.notSupported;
  }

  return ValidationType.unknown;
};
