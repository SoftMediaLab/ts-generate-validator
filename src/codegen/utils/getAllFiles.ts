import * as fs from 'fs';

export const getAllFiles = (path: string): string[] => {
  const tsRegExp = /.+\.ts$/;
  const returnFiles: string[] = [];

  const files = fs.readdirSync(path);
  files.forEach((file) => {
    const endPath = `${path}/${file}`;
    if (fs.statSync(endPath).isDirectory()) {
      const subFiles = getAllFiles(endPath);
      returnFiles.push(...subFiles);
    } else {
      const matches = tsRegExp.exec(endPath);
      if (matches && matches.length > 0) {
        returnFiles.push(matches[0]);
      }
    }
  });
  return returnFiles;
};
