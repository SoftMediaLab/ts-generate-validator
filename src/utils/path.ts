export const getRootDir = (): string => {
  return process.cwd();
};

export const cutFileExt = (file: string): string => {
  return file.replace(/\.[^/.]+$/, '');
};

export const hasFileExt = (file: string): boolean => {
  return /\.[^/.]+$/.test(file);
};

export const normalizePath = (rawPath: string): string => {
  return rawPath.replace(/\\+/g, '/');
};
