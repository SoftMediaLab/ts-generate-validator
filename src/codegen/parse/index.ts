import { cutFileExt, hasFileExt, isPackagePath, normalizeFileExt, normalizePath } from './../../utils/path';
import { IgnoreValidation } from './../../decorators/index';
import { IssueError, ErrorInFile, outWarning } from './../utils/error';
import { ValidationType } from './../../validators/model';
import { buildClassesMetadata } from './classes';
import { parseStruct, ImportNode } from 'ts-file-parser';
import * as fs from 'fs';
import * as path from 'path';
import {
  InputFileMetadata,
  ClassFieldTypeMetadata,
  ClassFieldBasicTypeMetadata,
  CustomTypeEntry,
  EnumDictionary,
  ImportMetadata,
  InputFileDesc,
  ClassMetadata
} from './model';

export const parseInputFiles = (files: string[]): InputFileMetadata[] => {
  const inputFiles: InputFileDesc[] = files.map((f) => ({ path: f, parseClasses: true }));
  const metadata: InputFileMetadata[] = [];
  const customTypeEntries: CustomTypeEntry[] = [];
  const enumDictionary: EnumDictionary = {};

  do {
    const inputFileDesc = inputFiles.shift();

    if (!inputFileDesc) {
      break;
    }

    const inputFilePath = path.resolve(process.cwd(), normalizeFileExt(inputFileDesc.path));
    let content: string;
    try {
      content = fs.readFileSync(inputFilePath, 'utf-8');
    } catch (err) {
      outWarning(`Failed to read input file "${inputFilePath}" -> skip (${err.message})`);
      continue;
    }
    const structure = parseStruct(content, {}, inputFileDesc.path);

    if (structure.enumDeclarations.length) {
      enumDictionary[inputFileDesc.path] = structure.enumDeclarations.map((decl) => decl.name);
    }

    const imports = buildImportsMetadata(structure._imports);
    imports.forEach((imp) => {
      const isAlreadyParsed = metadata.some((m) => m.name === imp.absPath);
      if (inputFileDesc.parseClasses && !isAlreadyParsed && !isPackagePath(imp.absPath)) {
        inputFiles.push({ path: imp.absPath, parseClasses: false });
      }
    });

    const classes: ClassMetadata[] = [];
    if (inputFileDesc.parseClasses && structure.classes.length) {
      const classesMetadata = buildClassesMetadata(metadata.length, structure.classes, structure._imports);
      if (classesMetadata.classesForValidation.length) {
        customTypeEntries.push(...classesMetadata.customTypeEntries);
        classes.push(...classesMetadata.classesForValidation);
      }
    }

    metadata.push({
      name: structure.name,
      classes,
      imports,
      functions: structure.functions.map(({ name, isExport, isAsync }) => ({ name, isExported: isExport, isAsync }))
    });
  } while (inputFiles.length);

  resolveCustomTypes({ metadata, enumDictionary, customTypeEntries });
  validateNestedClasses({ metadata, customTypeEntries });

  return metadata;
};

export const resolveCustomTypes = ({
  metadata,
  enumDictionary,
  customTypeEntries
}: {
  metadata: InputFileMetadata[];
  enumDictionary: EnumDictionary;
  customTypeEntries: CustomTypeEntry[];
}): void => {
  customTypeEntries.forEach(({ fileIndex, classIndex, fieldIndex }) => {
    const fieldTypeMetadata = metadata[fileIndex].classes[classIndex].fields[fieldIndex].type;

    const resolveBasicType = (typeMetadata: ClassFieldBasicTypeMetadata): void => {
      if (typeMetadata.validationType !== ValidationType.unknown) {
        return;
      }

      // Trying to fix empty referencePath -> set to current file with model class
      if (!typeMetadata.referencePath) {
        typeMetadata.referencePath = metadata[fileIndex].name;
      }

      const { referencePath, name: typeName } = typeMetadata;
      if (!referencePath || !typeName) {
        typeMetadata.validationType = ValidationType.notSupported;
        return;
      }

      if (enumDictionary[referencePath]?.includes(typeName)) {
        typeMetadata.validationType = ValidationType.enum;
        return;
      }

      const metadataItemWithNestedClass = metadata.find(({ name: fileName, classes }) => {
        let fileNameMatched = true;

        const fileNameHasExt = hasFileExt(fileName);
        const refPathHasExt = hasFileExt(referencePath);

        if (fileNameHasExt === refPathHasExt) {
          fileNameMatched = fileName === referencePath;
        } else {
          fileNameMatched = cutFileExt(fileName) === cutFileExt(referencePath);
        }

        if (!fileNameMatched) {
          return false;
        }

        return Boolean(classes.find(({ name: className }) => className === typeName));
      });
      if (metadataItemWithNestedClass) {
        typeMetadata.validationType = ValidationType.nested;
        typeMetadata.referencePath = metadataItemWithNestedClass.name;
        return;
      }

      typeMetadata.validationType = ValidationType.notSupported;
    };

    const resolveAnyType = (typeMetadata: ClassFieldTypeMetadata): void => {
      if (typeMetadata.validationType === ValidationType.array) {
        return resolveAnyType(typeMetadata.arrayOf);
      }
      if (typeMetadata.validationType === ValidationType.union) {
        return typeMetadata.unionTypes.forEach(resolveAnyType);
      }
      return resolveBasicType(typeMetadata);
    };

    resolveAnyType(fieldTypeMetadata);
  });
};

export const validateNestedClasses = ({
  metadata,
  customTypeEntries
}: {
  metadata: InputFileMetadata[];
  customTypeEntries: CustomTypeEntry[];
}): void => {
  customTypeEntries.forEach(({ fileIndex, classIndex, fieldIndex }) => {
    const baseClassName = metadata[fileIndex].classes[classIndex].name;
    const baseClassPath = metadata[fileIndex].name;
    const fieldTypeMetadata = metadata[fileIndex].classes[classIndex].fields[fieldIndex].type;

    const checkBasicType = (typeMetadata: ClassFieldBasicTypeMetadata): void => {
      if (typeMetadata.validationType !== ValidationType.nested) {
        return;
      }

      const checkNestedClasses = (className: string, classPath: string) => {
        if (className === baseClassName && classPath === baseClassPath) {
          const fieldName = metadata[fileIndex].classes[classIndex].fields[fieldIndex].name;
          throw new ErrorInFile(
            `Class "${baseClassName}" has circular dependency in field "${fieldName}" with type "${typeMetadata.name}". Ð¡ircular dependencies are not allowed because they lead to infinite validation. Change field type or add "@${IgnoreValidation.name}" decorator.`,
            baseClassPath
          );
        }

        const _fileMetadata = metadata.find(({ name: filePath }) => filePath === classPath);
        if (!_fileMetadata) {
          throw new IssueError(
            `Referenced file metadata not found for nested class "${className}" (referenced file path: "${classPath}").`
          );
        }

        const _classMetadata = _fileMetadata.classes.find(({ name }) => name === className);
        if (!_classMetadata) {
          throw new IssueError(`Metadata not found for nested class "${className}" in file metadata "${classPath}").`);
        }

        _classMetadata.fields.forEach(({ type: _fieldTypeMetadata }) => {
          const _checkBasicType = (_typeMetadata: ClassFieldBasicTypeMetadata): void => {
            const { name: _name, validationType: _validationType, referencePath: _referencePath } = _typeMetadata;

            if (_validationType === ValidationType.nested && _name && _referencePath) {
              checkNestedClasses(_name, _referencePath);
            }
          };

          const _checkAnyType = (_typeMetadata: ClassFieldTypeMetadata): void => {
            if (_typeMetadata.validationType === ValidationType.array) {
              return _checkAnyType(_typeMetadata.arrayOf);
            }
            if (_typeMetadata.validationType === ValidationType.union) {
              return _typeMetadata.unionTypes.forEach(_checkAnyType);
            }
            return _checkBasicType(_typeMetadata);
          };

          _checkAnyType(_fieldTypeMetadata);
        });
      };

      if (typeMetadata.name && typeMetadata.referencePath) {
        checkNestedClasses(typeMetadata.name, typeMetadata.referencePath);
      }
    };

    const checkAnyType = (typeMetadata: ClassFieldTypeMetadata): void => {
      if (typeMetadata.validationType === ValidationType.array) {
        return checkAnyType(typeMetadata.arrayOf);
      }
      if (typeMetadata.validationType === ValidationType.union) {
        return typeMetadata.unionTypes.forEach(checkAnyType);
      }
      return checkBasicType(typeMetadata);
    };

    checkAnyType(fieldTypeMetadata);
  });
};

const buildImportsMetadata = (nodes: ImportNode[]): ImportMetadata[] => {
  return nodes.map(({ clauses, absPathString }) => {
    return {
      clauses,
      absPath: normalizePath(path.relative(process.cwd(), path.resolve(absPathString)))
    };
  });
};
