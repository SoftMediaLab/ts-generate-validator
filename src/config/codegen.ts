import { outWarning } from './../codegen/utils/error';
import { getSanitizedObjectCopy, isObject } from './../utils/object';
import { getEnumValues } from './../utils/enum';
import { CodegenConfig, PartialCodegenConfig, SeverityLevel, codegenConfigKeys } from './model';
import { defaultCodegenConfig } from './default';
import { mergeDeep } from './../utils/deepValue';
import * as fs from 'fs';
import * as path from 'path';

export const configFileName = 'ts-generate-validator-config.json';

const getConfigFromFile = (): Record<string, any> => {
  const configFilePath = path.resolve(process.cwd(), configFileName);

  if (!fs.existsSync(configFilePath)) {
    return {};
  }
  try {
    const parsedData = JSON.parse(fs.readFileSync(configFilePath, 'utf8'));

    if (!isObject(parsedData)) {
      throw new Error();
    }

    return parsedData;
  } catch (err) {
    outWarning(
      `Error when reading config from file "${configFileName}". Make sure the file has correct syntax. Using default config...`
    );
    return {};
  }
};

export const getCodegenConfig = (): CodegenConfig => {
  const configFromFile = getConfigFromFile();
  const sanitizedConfig = getSanitizedObjectCopy<PartialCodegenConfig>(configFromFile, codegenConfigKeys);

  if (sanitizedConfig.hasOwnProperty('inputPath')) {
    if (typeof sanitizedConfig.inputPath !== 'string') {
      outWarning(
        `File "${configFileName}" has wrong configuration. Property "inputPath" is not a valid path. Using default value "${defaultCodegenConfig.inputPath}".`
      );
      delete sanitizedConfig.inputPath;
    }
  }

  if (sanitizedConfig.hasOwnProperty('outputPath')) {
    if (typeof sanitizedConfig.outputPath !== 'string') {
      outWarning(
        `File "${configFileName}" has wrong configuration. Property "outputPath" is not a valid path. Using default value "${defaultCodegenConfig.outputPath}".`
      );
      delete sanitizedConfig.inputPath;
    }
  }

  if (sanitizedConfig.hasOwnProperty('unknownPropertySeverityLevel')) {
    const possibleValues = getEnumValues(SeverityLevel);
    if (!possibleValues.includes(String(sanitizedConfig.unknownPropertySeverityLevel))) {
      outWarning(
        `File "${configFileName}" has wrong configuration. Property "unknownPropertySeverityLevel" must be one of "${possibleValues.join(
          ', '
        )}", but recieved "${sanitizedConfig.unknownPropertySeverityLevel}". Using default value "${
          defaultCodegenConfig.unknownPropertySeverityLevel
        }".`
      );
      delete sanitizedConfig.unknownPropertySeverityLevel;
    }
  }

  return mergeDeep({}, defaultCodegenConfig, sanitizedConfig) as CodegenConfig;
};
