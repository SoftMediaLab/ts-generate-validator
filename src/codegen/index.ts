import { parseInputFiles } from './parse';
import { getAllFiles } from './utils/getAllFiles';
import { getCodegenConfig } from './../config/codegen';

export const createValidators = (): void => {
  const config = getCodegenConfig();
  const inputFiles = getAllFiles(config.inputPath);
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  const inputFilesMetadata = parseInputFiles(inputFiles);

  // TODO: implement
};
