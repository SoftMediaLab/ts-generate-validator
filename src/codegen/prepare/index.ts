import { allowedFileExt } from './../model';
import { normalizePath, cutFileExt, isPackagePath } from './../../utils/path';
import { buildValidationFromClassMetadata } from './validation';
import { PreparedDataItem, PreparedValidation, PreparedImportMap, PreparedImport } from './model';
import * as path from 'path';
import { InputFileMetadata } from '../parse/model';
import { CodegenConfig } from '../../config/model';
import * as pkg from '../../../package.json';

export const prepareDataForRender = (
  inputFilesMetadata: InputFileMetadata[],
  config: CodegenConfig
): PreparedDataItem[] => {
  const preparedData: PreparedDataItem[] = [];

  inputFilesMetadata.forEach((metadata) => {
    const { name, classes, imports: inputFileImportsMetadata, functions: inputFileFunctionsMetadata } = metadata;
    const inputFilePath = cutFileExt(name);

    const filePath = buildOutputFilePath({ inputFileName: name, config });
    const fileName = buildOutputFileName(name);
    const filePathAbs = `${filePath}/${fileName}`;

    const importMap = buildBaseImportMap();
    const handleImportAdd = (targetPath: string, clause: string, isPackageName?: boolean): void => {
      const importPathAbs = path.relative(process.cwd(), targetPath);

      if (filePathAbs === importPathAbs) {
        return;
      }

      isPackageName = isPackageName ?? isPackagePath(targetPath);
      const importPath = isPackageName ? targetPath : normalizeImportPathForFile(filePath, targetPath);

      if (!importMap[importPath]) {
        importMap[importPath] = {};
      }

      importMap[importPath][clause] = true;
    };

    const validations: PreparedValidation[] = [];

    classes.forEach((cls) => {
      const validation = buildValidationFromClassMetadata({
        cls,
        clsFileName: name,
        addImport: handleImportAdd,
        inputFileImportsMetadata,
        inputFileFunctionsMetadata,
        inputFilePath,
        config
      });
      if (validation) {
        handleImportAdd(path.resolve(name), cls.name, false);
        validations.push(validation);
      }
    });

    const imports = buildImportsFromMap(importMap);

    if (validations.length) {
      preparedData.push({
        filePath,
        fileName,
        imports,
        validations
      });
    }
  });

  return preparedData;
};

const normalizeImportPathForFile = (filePath: string, importPath: string): string => {
  let resultPath = normalizePath(path.relative(filePath, importPath));
  if (!resultPath.startsWith('./')) {
    resultPath = `./${resultPath}`;
  }
  if (resultPath.match(new RegExp(`\\.(${allowedFileExt.join('|')})$`))) {
    resultPath = cutFileExt(resultPath);
  }
  return resultPath;
};

export const buildOutputFilePath = ({ config }: { inputFileName: string; config: CodegenConfig }): string => {
  let result = path.resolve(process.cwd(), config.outputPath);
  result = path.relative(process.cwd(), result);
  return normalizePath(result);
};

export const buildOutputFileName = (inputFileName: string): string => {
  const pathToInputFileFromRoot = path.relative(process.cwd(), inputFileName);
  return (
    cutFileExt(pathToInputFileFromRoot)
      .replace(/\s/g, '_')
      .replace(/[/\\]+/g, '.') + '.ts'
  );
};

const buildBaseImportMap = (): PreparedImportMap => {
  const map: PreparedImportMap = {};
  map[pkg.name] = {
    GeneratedValidation: true,
    ValidationConfig: true,
    ValidationError: true,
    ValidationException: true,
    Data: true,
    UserContext: true,
    getConfig: true,
    mergeDeep: true
  };
  return map;
};

const buildImportsFromMap = (map: PreparedImportMap): PreparedImport[] => {
  return Object.keys(map).map((path) => ({
    clauses: Object.keys(map[path])
      .filter((clause) => map[path][clause])
      .join(', '),
    path
  }));
};
