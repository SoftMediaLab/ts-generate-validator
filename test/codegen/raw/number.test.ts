import { prepareDataForRender } from './../../../src/codegen/prepare/index';
import { parseInputFiles } from '../../../src/codegen/parse/index';
import * as path from 'path';
import { getAllFiles } from '../../../src/codegen/utils/getAllFiles';
import { getCodegenConfig } from '../../../src/config/codegen';
import { deleteConfigFile, createConfigFile } from '../../utils/config';

describe('codegen/raw/number', () => {
  const prevRootDir = process.cwd();

  beforeAll(() => {
    const testRootDir = process.cwd() + '/test/mock/raw/number';
    process.chdir(testRootDir);
    createConfigFile({
      inputPath: '.'
    });
  });

  afterAll(() => {
    deleteConfigFile();
    process.chdir(prevRootDir);
  });

  test('parse & prepare', () => {
    const config = getCodegenConfig();
    const inputFiles = getAllFiles(path.resolve(config.inputPath));

    const inputFilesMetadata = parseInputFiles(inputFiles);
    expect(inputFilesMetadata).toMatchSnapshot('parseMetadata');

    // TODO: make separate tests for input files metadata

    // const dataForRender = prepareDataForRender(inputFilesMetadata, config);
    // expect(dataForRender).toMatchSnapshot('data for render');
  });
});
