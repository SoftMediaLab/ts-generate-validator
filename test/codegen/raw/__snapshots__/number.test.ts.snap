// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`codegen/raw/number parse & prepare: parseMetadata 1`] = `
Array [
  Object {
    "classes": Array [
      Object {
        "decorators": Array [
          Object {
            "arguments": Array [],
            "name": "Validation",
          },
        ],
        "fields": Array [
          Object {
            "decorators": Array [
              Object {
                "arguments": Array [
                  10,
                ],
                "name": "MinValidation",
              },
            ],
            "name": "someProperty",
            "optional": false,
            "type": Object {
              "name": "number",
              "validationType": "number",
            },
          },
        ],
        "name": "MinValidatorWithDefaultMessage",
      },
      Object {
        "decorators": Array [
          Object {
            "arguments": Array [],
            "name": "Validation",
          },
        ],
        "fields": Array [
          Object {
            "decorators": Array [
              Object {
                "arguments": Array [
                  10,
                ],
                "name": "MinValidation",
              },
            ],
            "name": "someProperty",
            "optional": false,
            "type": Object {
              "name": "string",
              "validationType": "string",
            },
          },
        ],
        "name": "MinValidatorOnWrongPropertyType",
      },
      Object {
        "decorators": Array [
          Object {
            "arguments": Array [],
            "name": "Validation",
          },
        ],
        "fields": Array [
          Object {
            "decorators": Array [
              Object {
                "arguments": Array [
                  10,
                  "min custom message",
                ],
                "name": "MinValidation",
              },
            ],
            "name": "someProperty",
            "optional": false,
            "type": Object {
              "name": "number",
              "validationType": "number",
            },
          },
        ],
        "name": "MinValidatorWithCustomMessage",
      },
      Object {
        "decorators": Array [
          Object {
            "arguments": Array [],
            "name": "Validation",
          },
        ],
        "fields": Array [
          Object {
            "decorators": Array [
              Object {
                "arguments": Array [
                  10,
                ],
                "name": "MaxValidation",
              },
            ],
            "name": "someProperty",
            "optional": false,
            "type": Object {
              "name": "number",
              "validationType": "number",
            },
          },
        ],
        "name": "MaxValidatorWithDefaultMessage",
      },
      Object {
        "decorators": Array [
          Object {
            "arguments": Array [],
            "name": "Validation",
          },
        ],
        "fields": Array [
          Object {
            "decorators": Array [
              Object {
                "arguments": Array [
                  10,
                ],
                "name": "MaxValidation",
              },
            ],
            "name": "someProperty",
            "optional": false,
            "type": Object {
              "name": "string",
              "validationType": "string",
            },
          },
        ],
        "name": "MaxValidatorOnWrongPropertyType",
      },
      Object {
        "decorators": Array [
          Object {
            "arguments": Array [],
            "name": "Validation",
          },
        ],
        "fields": Array [
          Object {
            "decorators": Array [
              Object {
                "arguments": Array [
                  10,
                  "max custom message",
                ],
                "name": "MaxValidation",
              },
            ],
            "name": "someProperty",
            "optional": false,
            "type": Object {
              "name": "number",
              "validationType": "number",
            },
          },
        ],
        "name": "MaxValidatorWithCustomMessage",
      },
      Object {
        "decorators": Array [
          Object {
            "arguments": Array [],
            "name": "Validation",
          },
        ],
        "fields": Array [
          Object {
            "decorators": Array [
              Object {
                "arguments": Array [],
                "name": "NegativeValidation",
              },
            ],
            "name": "someProperty",
            "optional": false,
            "type": Object {
              "name": "number",
              "validationType": "number",
            },
          },
        ],
        "name": "NegativeValidatorWithDefaultMessage",
      },
      Object {
        "decorators": Array [
          Object {
            "arguments": Array [],
            "name": "Validation",
          },
        ],
        "fields": Array [
          Object {
            "decorators": Array [
              Object {
                "arguments": Array [],
                "name": "NegativeValidation",
              },
            ],
            "name": "someProperty",
            "optional": false,
            "type": Object {
              "name": "string",
              "validationType": "string",
            },
          },
        ],
        "name": "NegativeValidatorOnWrongPropertyType",
      },
      Object {
        "decorators": Array [
          Object {
            "arguments": Array [],
            "name": "Validation",
          },
        ],
        "fields": Array [
          Object {
            "decorators": Array [
              Object {
                "arguments": Array [
                  "negative custom message",
                ],
                "name": "NegativeValidation",
              },
            ],
            "name": "someProperty",
            "optional": false,
            "type": Object {
              "name": "number",
              "validationType": "number",
            },
          },
        ],
        "name": "NegativeValidatorWithCustomMessage",
      },
      Object {
        "decorators": Array [
          Object {
            "arguments": Array [],
            "name": "Validation",
          },
        ],
        "fields": Array [
          Object {
            "decorators": Array [
              Object {
                "arguments": Array [],
                "name": "PositiveValidation",
              },
            ],
            "name": "someProperty",
            "optional": false,
            "type": Object {
              "name": "number",
              "validationType": "number",
            },
          },
        ],
        "name": "PositiveValidatorWithDefaultMessage",
      },
      Object {
        "decorators": Array [
          Object {
            "arguments": Array [],
            "name": "Validation",
          },
        ],
        "fields": Array [
          Object {
            "decorators": Array [
              Object {
                "arguments": Array [],
                "name": "PositiveValidation",
              },
            ],
            "name": "someProperty",
            "optional": false,
            "type": Object {
              "name": "string",
              "validationType": "string",
            },
          },
        ],
        "name": "PositiveValidatorOnWrongPropertyType",
      },
      Object {
        "decorators": Array [
          Object {
            "arguments": Array [],
            "name": "Validation",
          },
        ],
        "fields": Array [
          Object {
            "decorators": Array [
              Object {
                "arguments": Array [
                  "positive custom message",
                ],
                "name": "PositiveValidation",
              },
            ],
            "name": "someProperty",
            "optional": false,
            "type": Object {
              "name": "number",
              "validationType": "number",
            },
          },
        ],
        "name": "PositiveValidatorWithCustomMessage",
      },
      Object {
        "decorators": Array [
          Object {
            "arguments": Array [],
            "name": "Validation",
          },
        ],
        "fields": Array [
          Object {
            "decorators": Array [
              Object {
                "arguments": Array [],
                "name": "IntegerValidation",
              },
            ],
            "name": "someProperty",
            "optional": false,
            "type": Object {
              "name": "number",
              "validationType": "number",
            },
          },
        ],
        "name": "IntegerValidatorWithDefaultMessage",
      },
      Object {
        "decorators": Array [
          Object {
            "arguments": Array [],
            "name": "Validation",
          },
        ],
        "fields": Array [
          Object {
            "decorators": Array [
              Object {
                "arguments": Array [],
                "name": "IntegerValidation",
              },
            ],
            "name": "someProperty",
            "optional": false,
            "type": Object {
              "name": "string",
              "validationType": "string",
            },
          },
        ],
        "name": "IntegerValidatorOnWrongPropertyType",
      },
      Object {
        "decorators": Array [
          Object {
            "arguments": Array [],
            "name": "Validation",
          },
        ],
        "fields": Array [
          Object {
            "decorators": Array [
              Object {
                "arguments": Array [
                  "integer custom message",
                ],
                "name": "IntegerValidation",
              },
            ],
            "name": "someProperty",
            "optional": false,
            "type": Object {
              "name": "number",
              "validationType": "number",
            },
          },
        ],
        "name": "IntegerValidatorWithCustomMessage",
      },
      Object {
        "decorators": Array [
          Object {
            "arguments": Array [],
            "name": "Validation",
          },
        ],
        "fields": Array [
          Object {
            "decorators": Array [
              Object {
                "arguments": Array [],
                "name": "FloatValidation",
              },
            ],
            "name": "someProperty",
            "optional": false,
            "type": Object {
              "name": "number",
              "validationType": "number",
            },
          },
        ],
        "name": "FloatValidatorWithDefaultMessage",
      },
      Object {
        "decorators": Array [
          Object {
            "arguments": Array [],
            "name": "Validation",
          },
        ],
        "fields": Array [
          Object {
            "decorators": Array [
              Object {
                "arguments": Array [],
                "name": "FloatValidation",
              },
            ],
            "name": "someProperty",
            "optional": false,
            "type": Object {
              "name": "string",
              "validationType": "string",
            },
          },
        ],
        "name": "FloatValidatorOnWrongPropertyType",
      },
      Object {
        "decorators": Array [
          Object {
            "arguments": Array [],
            "name": "Validation",
          },
        ],
        "fields": Array [
          Object {
            "decorators": Array [
              Object {
                "arguments": Array [
                  "float custom message",
                ],
                "name": "FloatValidation",
              },
            ],
            "name": "someProperty",
            "optional": false,
            "type": Object {
              "name": "number",
              "validationType": "number",
            },
          },
        ],
        "name": "FloatValidatorWithCustomMessage",
      },
      Object {
        "decorators": Array [
          Object {
            "arguments": Array [],
            "name": "Validation",
          },
        ],
        "fields": Array [
          Object {
            "decorators": Array [
              Object {
                "arguments": Array [
                  "someOtherProperty",
                ],
                "name": "LessThanValidation",
              },
            ],
            "name": "someProperty",
            "optional": false,
            "type": Object {
              "name": "number",
              "validationType": "number",
            },
          },
          Object {
            "decorators": Array [
              Object {
                "arguments": Array [
                  "someProperty",
                ],
                "name": "MoreThanValidation",
              },
            ],
            "name": "someOtherProperty",
            "optional": false,
            "type": Object {
              "name": "number",
              "validationType": "number",
            },
          },
        ],
        "name": "LessThanMoreThanValidatorWithDefaultMessage",
      },
      Object {
        "decorators": Array [
          Object {
            "arguments": Array [],
            "name": "Validation",
          },
        ],
        "fields": Array [
          Object {
            "decorators": Array [
              Object {
                "arguments": Array [
                  "someOtherProperty",
                ],
                "name": "LessThanValidation",
              },
            ],
            "name": "someProperty",
            "optional": false,
            "type": Object {
              "name": "string",
              "validationType": "string",
            },
          },
          Object {
            "decorators": Array [
              Object {
                "arguments": Array [
                  "someProperty",
                ],
                "name": "MoreThanValidation",
              },
            ],
            "name": "someOtherProperty",
            "optional": false,
            "type": Object {
              "name": "string",
              "validationType": "string",
            },
          },
        ],
        "name": "LessThanMoreThanValidatorOnWrongPropertyType",
      },
      Object {
        "decorators": Array [
          Object {
            "arguments": Array [],
            "name": "Validation",
          },
        ],
        "fields": Array [
          Object {
            "decorators": Array [
              Object {
                "arguments": Array [
                  "someOtherProperty",
                  "lessThan custom message",
                ],
                "name": "LessThanValidation",
              },
            ],
            "name": "someProperty",
            "optional": false,
            "type": Object {
              "name": "number",
              "validationType": "number",
            },
          },
          Object {
            "decorators": Array [
              Object {
                "arguments": Array [
                  "someProperty",
                  "moreThan custom message",
                ],
                "name": "MoreThanValidation",
              },
            ],
            "name": "someOtherProperty",
            "optional": false,
            "type": Object {
              "name": "number",
              "validationType": "number",
            },
          },
        ],
        "name": "LessThanValidatorWithCustomMessage",
      },
    ],
    "name": "model.ts",
  },
]
`;
